### PR TITLE
Grafana alerting integrated with Pagerduty

### LINKED ISSUE #6291: Alerting: Integration with PagerDuty
- **I'm submitting a ...**
- [ ] Bug report
- [x] Feature request
- [ ] Question / Support request: **Please do not** open a github issue. [Support Options](http://grafana.org/support/)

Would love to integrated Alerting with xMatters or PagerDuty.

### PR DESCRIPTION
**Pull request by utkarshcmu**
Resolves #6291 

Screenshot:

screen shot 2016-11-02 at 1 30 37 pm

### CODE REVIEW

**Comment by utkarshcmu** [comment]
PR ready for review/merge!

File: pkg/metrics/metrics.go
Comment (bergquist) [inline pkg/metrics/metrics.go]: Seems like a tabs vs space issue. What editor are you using? Any chance you can find a plugin for editorconfig? https://github.com/grafana/grafana/blob/master/.editorconfig http://editorconfig.org/

File: pkg/services/alerting/notifiers/pagerduty.go
Comment (bergquist) [inline pkg/services/alerting/notifiers/pagerduty.go]: I've never integrated with pagerduty. But it shouldn't it send notifications for ok as well?

Comment (utkarshcmu) [inline pkg/services/alerting/notifiers/pagerduty.go]: Pagerduty can handle alert acknowledgement and resolution on its own. Alerts just need to be triggered from an external source.

File: pkg/metrics/metrics.go
Comment (bergquist) [inline pkg/metrics/metrics.go]: ofc! Sorry about that :)

**Comment by bergquist** [comment]
This might take some time to review since I have to setup a pagerduty account for testing.

File: pkg/metrics/metrics.go
Comment (utkarshcmu) [inline pkg/metrics/metrics.go]: No worries. :)

**Comment by utkarshcmu** [comment]
Blocked by #6420

File: pkg/services/alerting/notifiers/pagerduty.go
Comment (arodland) [inline pkg/services/alerting/notifiers/pagerduty.go]: @utkarshcmu sending OK is better if possible. There's a reason why they provide it in the API. "Your monitoring tools should send PagerDuty a `resolve` event when the problem that caused the original trigger is fixed."

Comment (bergquist) [inline pkg/services/alerting/notifiers/pagerduty.go]: Perhaps it should be optional. I guess some organisation will require a human to set it as resolved.

Comment (utkarshcmu) [inline pkg/services/alerting/notifiers/pagerduty.go]: I will keep it optional in the config

Comment (utkarshcmu) [inline pkg/services/alerting/notifiers/pagerduty.go]: @arodland Storing event Id of the incident will take some more effort in PagerDuty support. Will add the feature of resolving in the next version of the support. For now when we merge this PR, Grafana will be able to Trigger an incident in PagerDuty and from there on PagerDuty can take care of the incident.

**Comment by utkarshcmu** [comment]
Ready for Review/Merge.

**Comment by utkarshcmu** [comment]
Blocked on #6444 now!

**Comment by sofixa** [comment]
Just a feature request, could you make the PagerDuty url optional for the people running PagerDuty-compatible software(like OpenDuty)?

**Comment by utkarshcmu** [comment]
@sofixa Integration Key field is mandatory. There is no url field for PagerDuty settings.

**Comment by sofixa** [comment]
@utkarshcmu ,  i meant, could you add it as a field/parameter, so that people who use compatible software can use your alerter out-of the box?

**Comment by bergquist** [comment]
Since https://github.com/grafana/grafana/pull/6489 have been merged we dont have a state for execution_alerts anymore. So please remove that filter. I will review this once that is fixed so hopefully this can be merge before we release 4.0 beta.

**Review by bergquist — CHANGES_REQUESTED** [review]
Since #6489 have been merged we dont have a state for execution_alerts anymore. So please remove that filter. I will review this once that is fixed so hopefully this can be merge before we release 4.0 beta.

**Comment by utkarshcmu** [comment]
@bergquist Great work on #6489 , I have removed the ExecError state. Please review/merge. Thanks!

**Comment by bergquist** [comment]
Id like an option for auto resolving the pagerduty status when going back to ok. I can understand organizations that don't want to auto resolve alerts but we some do. Including us internally :)

**Comment by utkarshcmu** [comment]
@bergquist For auto resolving the pagerduty incident , I will need to store the eventId in the db. Let me see where I can do that in the db.

**Comment by bergquist** [comment]
@utkarshcmu oh. Lets postpone that for now.

**Comment by bergquist** [comment]
Great work!

**Comment by utkarshcmu** [comment]
@sofixa Created an issue to track support for OpenDuty - #6504

**Comment by playqdrew** [comment]
@utkarshcmu  -- Thanks for the work on this! Very nice. 

I am having an issue when alerts are sent to PagerDuty versus Slack. Some information seems to be missing. Here is an example in Slack:

image

Versus what shows for the same alerts in PagerDuty:

{
  "client": "Grafana",
  "client_url": "https://xxx/dashboard/db/prd-tg-overview?fullscreen&edit&tab=alert&panelId=3&orgId=1",
  "description": "PRD TG-Vital Request Timings - ",
  "event_type": "trigger",
  "incident_key": "alertId-16",
  "service_key": "xxx"
}

Notice how the whole description about how this was an influxdb at capacity error is missing from PagerDuty?

**Pull request merged**
